hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/NioAsyncSocketBuilder.java

```diff
@@ -34,15 +37,35 @@
  * A {@link AsyncSocketBuilder} specific to the {@link NioAsyncSocket}.
  */
 public class NioAsyncSocketBuilder implements AsyncSocketBuilder {
+
     static final int DEFAULT_WRITE_QUEUE_CAPACITY = 2 << 16;
 
+    private static final Constructor<AsyncSocket> TLS_NIO_ASYNC_SOCKET_CONSTRUCTOR;
+
+    private static final String TLS_NIO_ASYNC_SOCKET_CLASS_NAME = "com.hazelcast.internal.tpcengine.nio.TlsNioAsyncSocket";
+
+    static {
+        Constructor<AsyncSocket> tlsNioAsyncSocketConstructor = null;
+        try {
+            Class<?> clazz = NioAsyncSocketBuilder.class.getClassLoader().loadClass(TLS_NIO_ASYNC_SOCKET_CLASS_NAME);
```

Comment:
Can you explain what was the reason to use reflection and load constructor at runtime? Could you not use the NodeExtension interface and add a method there to get this class?

Reply:
@pveentjer would you mind pitching in?

Reply:
The TPC engine is properly modularized (it is in its own maven module) and hence it doesn't have access to the NodeExtension. It would also be a false dependency since the TPC engine should not need to care about NodeEngine.

Reply:
One of the reasons the client needs to have the full cluster code is due to lack of modularization. E.g. since classic networking is part of the main hazelcast module and client uses classic networking, the client needs to rely on the main hazelcast module. There is a similar problem with e.g. logger that caries a dependency to cluster specific code.